section 3

aleno4ka1998 opened the pull request:
for and while

aleno4ka1998 committed:
Сумма квадратов

По данному натуральному n вычислите сумму 12+22+...+n2.

Входные данные
Вводится единственное натуральное число n, не превосходящее 100

Выходные данные
Необходимо вывести  вычисленную сумму.

aleno4ka1998 committed:
Факториал

Вычислите N! ("эн-факториал") – произведение всех натуральных чисел от 1 до N ( N!=1∙2∙3∙…∙ N ).

Входные данные
Вводится единственное число N – натуральное, не превосходит 12.

Выходные данные
Выведите полученное значение N!

aleno4ka1998 committed:
Степень

Напишите программу, вычисляющую 2 N.

Входные данные
Вводится целое неотрицательное число N, которое не превосходит 30.

Выходные данные
Выведите число 2 N.

aleno4ka1998 committed:
Число сочетаний

По данным натуральным n и k вычислите значение Ckn=n!/k!(n−k)! (число сочетаний из n элементов по k).

Входные данные
Вводятся 2 числа - n и k (n,k≤30).

Выходные данные
Необходимо вывести  значение Ckn.

aleno4ka1998 committed:
Геометрическая прогрессия

По данному действительному числу a и натуральному n вычислите сумму 1+a+a2+...+an, не используя формулу суммы геометрической прогрессии. Время работы программы должно быть пропорционально n.

Входные данные
Вводятся 2 числа - a и n.

Выходные данные
Необходимо вывести  значение суммы.

aleno4ka1998 committed:
Сумма - 1

По данному числу n вычислите сумму 1+1/22+1/32+...+1/n2.

Входные данные
Вводится одно число n, не превосходящее 100000.

Выходные данные
Необходимо вывести  значение суммы.

aleno4ka1998 committed:
Сумма - 2

По данному числу n вычислите сумму 4(1-1/3+1/5-1/7+...+(-1)n/(2n+1)).

Входные данные
Вводится одно число n, не превосходящее 100000.

Выходные данные
Необходимо вывести  значение выражения.

aleno4ka1998 committed:
Сумма степеней

Входные данные
Вводится натуральное число N, которое не превосходит 30.

Выходные данные
Вычислите 1+2+22+23+…+2 N.

thelightningseas commented on section3/combinations.py:
```diff
@@ -0,0 +1,12 @@
+n = int(input())
```
тут ты 3 раза делаешь очень похожие вещи. из-за этого во-первых тратится врмея на то чтобы писать одно и то же, во вторых копируешь свои же ошибки, в третьих выглядит не так читаемо. лучше в таких ситуациях пользоваться функциями: 

```
def factorial(n):
    res = 1
    for i in range(1, n + 1):
        res *= i
    return res


n = int(input())
k = int(input())

print(factorial(n) / (factorial(k) * factorial(n- k)))
```
```

thelightningseas replied:
еще можно написать функцию факториала покрасивее - через рекурсию

```
def factorial(n):
    if n == 1:
        return 1
    return n * factorial(n - 1)
````

thelightningseas commented on section3/exponent.py:
```diff
@@ -0,0 +1,5 @@
+total = 2
+for i in range(1):
+    total **= n
+print(total)
```
тут конечно ответ верный, но думаю, подразумевалось что-то вроде:

```
for i in range(n): total *= 2
```

а то можно было бы и просто ```print(2 ** int(input()))``` написать)

aleno4ka1998 committed:
1/0!+1/1!+1/2!+...

По данному натуральному числу N найдите сумму чисел 1+1/1!+1/2!+1/3!+...+1/N!. Количество действий должно быть пропорционально N.

Входные данные
Задано единственное число N

Выходные данные
Необходимо вывести  результат вычисления в виде действительного числа c точностью до 5 знаков после запятой.

aleno4ka1998 committed:
Четные числа

Входные данные
Вводятся целые числа a и b. Гарантируется, что a не превосходит b

Выходные данные
Выведите (через пробел) все четные числа от a до b (включительно).

aleno4ka1998 committed:
Остаток

Входные данные
Вводятся 4 числа: a, b, c и d.

Выходные данные
Выведите все числа на отрезке от a до b, дающие остаток c при делении на d. Если таких чисел не существует, то ничего выводить не нужно.

aleno4ka1998 committed:
Квадраты

Входные данные
Вводятся целые числа a и b. Гарантируется, что a не превосходит b.

Выходные данные
Выведите все числа на отрезке от a до b, являющиеся полными квадратами. Если таких чисел нет, то ничего выводить не нужно.

aleno4ka1998 committed:
Минимальный делитель

Найдите самый маленький натуральный делитель числа x, отличный от 1 (2 <= x <= 30000).

Входные данные
Вводится натуральное число x.

Выходные данные
Выведите наименьший делитель числа x, отличный от 1.

thelightningseas commented on section3/even_numbers.py:
```diff
@@ -0,0 +1,5 @@
+a = int(input())
+b = int(input())
+for i in range(a, b + 1):
```
тут возможно хотели рассказать что у range есть еще параметер ```step```
четные числа идут через одно, поэтому можно сделать ```range(a, b + 1, 2)```
только надо учесть что если ```a``` нечетное, то начать надо с ```a + 1```

thelightningseas replied:
например можно начинать с ```a + a % 2```

thelightningseas replied:
```
a = int(input())
b = int(input())

for i in range(a + a % 2, b + 1, 2):
    print(i, end=" ")
```

thelightningseas commented on section3/minimum_divisor.py:
```diff
@@ -0,0 +1,6 @@
+x = int(input())
+y = 0
+for i in range(x, 1, -1):
+    if x % i == 0:
```
тут хорошо использовать ```break``` - штука, которая останвливает цикл досрочно

```
x = int(input())

for i in range(2, x + 1):
    if x % i == 0:
        print(i)
        break
```

aleno4ka1998 committed:
Делители числа

Выведите все натуральные делители числа x в порядке возрастания (включая 1 и само число).

Входные данные
Вводится натуральное число x

Выходные данные
Выведите все делители числа x

aleno4ka1998 committed:
Количество делителей

Подсчитайте количество натуральных делителей числа x (включая 1 и само число; x≤2∗109).

Входные данные
Вводится натуральное число x.

Выходные данные
Выведите единственное число - количество делителей числа x

aleno4ka1998 committed:
Сумма ста

Вычислите сумму данных 100 натуральных чисел.

Входные данные
Вводятся 100 чисел, сумму которых необходимо посчитать.

Выходные данные
Программа должна вывести единственное число - полученную сумму.

aleno4ka1998 committed:
Сумма чисел

Вычислите сумму данных N натуральных чисел.

Входные данные
Вводится число N, а затем N чисел, сумму которых необходимо вычислить.

Выходные данные
Выведите единственное число - сумму введенных чисел.

aleno4ka1998 committed:
Нули

Входные данные
Вводится число N, а затем N чисел.

Выходные данные
Подсчитайте и выведите, сколько среди данных N чисел нулей.

aleno4ka1998 committed:
Подсчёт чисел

Подсчитайте, сколько среди данных N чисел нулей, положительных чисел, отрицательных чисел.

Входные данные
Вводится число N, а затем N целых чисел.

Выходные данные
Необходимо вывести сначала число нулей, затем число положительных и отрицательных чисел.

aleno4ka1998 committed:
ноль или не ноль

Проверьте, есть ли среди данных N чисел нули.

Входные данные
Вводится число N, а затем N чисел.

Выходные данные
Выведите YES, если среди введенных чисел есть хотя бы один нуль, или NO в противном случае.

thelightningseas commented on section3/counting_numbers.py:
```diff
@@ -0,0 +1,13 @@
+null = 0
+positive = 0
+negative = 0
+for i in range(1, n + 1):
```
зачем тут 1, n + 1? можно же просто n

еще принято не создавать переменную, если не используешь ее
```for _ in range(n)```

thelightningseas commented on section3/null_or_not.py:
```diff
@@ -0,0 +1,10 @@
+        nulls += 1
+if nulls == 0:
+    print('NO')
+else:
```
не надо делать юесполезную работу. если не нужно считать, сколько нулей, то не считай

```
n = int(input())
has_nulls = False

for i in range(1, n + 1):
    x = int(input())
    if x == 0:
        has_nulls = True

if has_nulls:
    print('NO')
else:
    print('YES')
```

thelightningseas replied:
ну и с неймингом. наверное стоит ```nulls -> zeroes```

thelightningseas commented on section3/square.py:
```diff
@@ -0,0 +1,5 @@
+a = int(input())
+b = int(input())
+for i in range(a, b + 1):
```
лучше так
```
from math import sqrt

int(sqrt(x)) ** 2 == x
```

thelightningseas commented on section3/sum_of_exponent.py:
```diff
@@ -0,0 +1,8 @@
+n = int(input())
+total = 1
+total1 = 1
```
было бы понятнее, если бы внутренний цикл был вынесен в функцию
и ```total1``` не очень хорошее имя переменной

aleno4ka1998 committed:
Уровнение по возрастанию

Входные данные
Вводятся 4 числа: a, b, c и d.

Выходные данные
Найдите все целые решения уравнения ax3 + bx2 + cx + d = 0 на отрезке [0,1000] и выведите их в порядке возрастания.  Если на данном отрезке нет ни одного решения, то ничего выводить не нужно.

aleno4ka1998 committed:
Уравнение по убыванию

Входные данные
Вводятся 4 числа: a, b, c и d.

Выходные данные
Найдите все целые решения уравнения ax3 + bx2 + cx + d = 0 на отрезке [0,1000] и выведите их в порядке убывания. Если на данном отрезке нет ни одного решения, то ничего выводить не нужно.

aleno4ka1998 committed:
Количество решений

Входные данные
Вводятся 5 чисел: a, b, c, d и e.

Выходные данные
Найдите все целые решения уравнения ( ax3 + bx2 + cx + d ) / ( x - e ) = 0 на отрезке [0,1000] и выведите их количество.

aleno4ka1998 committed:
ГНЧЭ-1

"ГНЧЭ-1"  – сложное электронное устройство, выдающее каждую секунду очередное число последовательности 1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 5... Ввиду дороговизны электронных комплектующих вам поручено разработать эмулятор для этого устройства.

Входные данные
Дано количество секунд (от 1 до 1000000), которые работает генератор после включения.
